Add coverage tests graph_listener

Blast545 opened the pull request:
Checking with CI coverage, I think I have a coverage issue with my local environment at the moment:
* Linux coverage [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux_coverage&build=207)](http://ci.ros2.org/job/ci_linux_coverage/207/)

brawner commented on rclcpp/test/rclcpp/test_graph_listener.cpp:
```diff
@@ -0,0 +1,309 @@
+
+    // This dynamic cast is not necessary for the unittests, but instead is used to ensure
+    // the proper type is being tested and covered.
+    node_graph_ =
```
Since this isn't a test about rclcpp::node_interfaces::NodeGraph I think it's fine if `node_graph_` was just a `rclcpp::node_interfaces::NodeGraphInterface`. Then you wouldn't need the dynamic cast

Blast545 replied:
Ok, see 08415ca

brawner commented on rclcpp/test/rclcpp/test_graph_listener.cpp:
```diff
@@ -0,0 +1,309 @@
+      dynamic_cast<rclcpp::node_interfaces::NodeGraph *>(node_->get_node_graph_interface().get());
+    ASSERT_NE(nullptr, node_graph_);
+
+    using rclcpp::contexts::get_global_default_context;
```
I don't think you need this `using` because it's used only once.

Blast545 replied:
See bb23126

brawner commented on rclcpp/test/rclcpp/test_graph_listener.cpp:
```diff
@@ -0,0 +1,309 @@
+};
+
+/* Run base class init/shutdown */
+TEST_F(TestGraphListener, construction_and_destruction) {
```
Are there any more checks you can do on the default `NodeGraphListener` for this test fixture? For example, `is_shutdown()` and `has_node` seem like good checks to do on your test fixture

Blast545 replied:
I'll add some shutdown error tests I think I have missed. As for the `has_node` function is tested with the `add_node` and `remove_node`, do you want them to be with `construction_and_destruction` for some reason in particular?

brawner replied:
The reasoning being that you're checking the state of the graph_listener after construction. It's not about checking the functionality of `is_shutdown()` or `has_node()`, but more about the expected state of graph_listener.

Blast545 replied:
I added those tests with 85f6933

brawner commented on rclcpp/test/rclcpp/test_graph_listener.cpp:
```diff
@@ -0,0 +1,309 @@
+  auto mock = mocking_utils::patch_and_return(
+    "lib:rclcpp", rcl_guard_condition_init, RCL_RET_ERROR);
+
+  using rclcpp::contexts::get_global_default_context;
```
this is a duplicate of line 84

Blast545 replied:
Ups, see dae2f04

brawner commented on rclcpp/test/rclcpp/test_graph_listener.cpp:
```diff
@@ -0,0 +1,309 @@
+  RCLCPP_EXPECT_THROW_EQ(
+  {
+    auto graph_listener_error =
+    std::make_shared<rclcpp::graph_listener::GraphListener>(get_global_default_context());
```
You think you should be using `.reset()`. Otherwise, I think you're likely to get an unused variable warning for graph_listener_error

```
RCLCPP_EXPECT_THROW_EQ(
    std::make_shared<rclcpp::graph_listener::GraphListener>(get_global_default_context()).reset(),
    std::runtime_error("..."));
```

Blast545 replied:
See f13344a

brawner replied:
This one is still unresolved I think

Blast545 replied:
I never got the warning you told me that happened, I added it anyways with: 18a897f

brawner replied:
I've only seen Windows complain about unused variables inside `RCLCPP_EXPECT_THROW_EQ`.

brawner commented on rclcpp/test/rclcpp/test_graph_listener.cpp:
```diff
@@ -0,0 +1,309 @@
+  {
+    auto mock = mocking_utils::patch_and_return(
+      "lib:rclcpp", rcl_wait_set_init, RCL_RET_ERROR);
+    RCLCPP_EXPECT_THROW_EQ(
```
For this macro, you can use it like `EXPECT_THROW` and you only need curly braces if you're testing multiple lines of code. Fix here and below

Blast545 replied:
See f13344a

brawner commented on rclcpp/test/rclcpp/test_graph_listener.cpp:
```diff
@@ -0,0 +1,309 @@
+    context_to_destroy->init(0, nullptr);
+    auto graph_listener_error =
+      std::make_shared<rclcpp::graph_listener::GraphListener>(context_to_destroy);
+    // rclcpp::shutdown(context_to_destroy); Using this causes test to fail
```
This would cause two shutdown's correct, since you're already doing it in your test fixture?

Blast545 replied:
I totally forgot about that shutdown, you are right, remove this comment with f402201

brawner commented on rclcpp/test/rclcpp/test_graph_listener.cpp:
```diff
@@ -0,0 +1,309 @@
+// Tricky: patch on return the start method, to prepare the
+// loop thread to be run, but leave early with the failure thrown
+// Then, run the loop method to test it
+#define MOCK_WAIT_SET_INIT_ON_THREAD() \
```
This is cool. Does this need to be a macro, or could you just create a member function for your test fixture? Also, it might be good to have an additional test that just checks this case without additional mocks

Blast545 replied:
I used a macro because the code is mocked, and AFAIK mocking inside a function will erase the mock after leaving the scope of the function.

As for the the mocked function, is tested on its own as the first test of `error_start_graph_listener`, do you mean a test on its own for the macro itself?

brawner replied:
Oh, you are correct about the function scope issue. But that raises another question, why doesn't `mock_wait_set_init` fall out of scope in this do-while loop?

You might just have to take the inject_on_error call and duplicate in each block below. The expected exception can just be checked the first time

Blast545 replied:
The purpose of the mocking function is to initialize the graph_listener variables without actually starting it. I just realized the mock goes out of scope, but the graph_listener stays in the desired state. So I think I can change it to be a member function of the test fixture.

Blast545 replied:
Changed this with: 06b3645

brawner commented on rclcpp/test/rclcpp/test_graph_listener.cpp:
```diff
@@ -0,0 +1,309 @@
+  } while (0)
+
+/* Errors running graph protected methods */
+TEST_F(TestGraphListener, error_run_graph_listener) {
```
This test is quite long. I think it just makes sense to turn each scope block into its own test.

Blast545 replied:
See bb8d166

brawner commented on rclcpp/test/rclcpp/test_graph_listener.cpp:
```diff
@@ -40,15 +40,13 @@ class TestGraphListener : public ::testing::Test
-    // the proper type is being tested and covered.
-    node_graph_ =
-      dynamic_cast<rclcpp::node_interfaces::NodeGraph *>(node_->get_node_graph_interface().get());
+    node_graph_ = std::make_shared<rclcpp::node_interfaces::NodeGraph>(
```
I think this can simply just be:
```
node_graph_ = node_->get_node_graph_interface();
```

Blast545 replied:
I can't assign a `node_graph_interface` to a `node_graph`, am I missing something in your suggestion?

brawner replied:
I was also thinking you can change the type of node_graph_ to `rclcpp::node_interfaces::NodeGraphInterface`. It's the base type, we don't need to dynamic cast to the derived type for this test.

brawner replied:
So in this class, the last changes I am looking for would be to adjust the definition of `node_graph_`

```
rclcpp::node_interfaces::NodeGraphInterface::SharedPtr node_graph_
```
Then this line can just be:
```
 node_graph_ = node_->get_node_base_interface();
```

Blast545 replied:
Then how I make the method `node_graph()` returning `rclcpp::node_interfaces::NodeGraph *` ?

Blast545 replied:
Ok, see 6114d38

Blast545 commented:
CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=12491)](http://ci.ros2.org/job/ci_linux/12491/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7457)](http://ci.ros2.org/job/ci_linux-aarch64/7457/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10203)](http://ci.ros2.org/job/ci_osx/10203/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=12411)](http://ci.ros2.org/job/ci_windows/12411/)

Blast545 commented:
There a couple of places I wasn't sure how to tackle, but so far I think I've addressed all your comments, let me know what you think @brawner 

CI_linux coverage job seems not to be working as reference for coverage.

brawner commented:
It looks like this PR fully covers the graph_listener.hpp and adds 21 lines of coverage to graph_listener.cpp. Are you expecting more?

Blast545 commented:
@brawner I wanted to tackle as well this section that wasn't covered: https://github.com/ros2/rclcpp/blob/148d295416dba3814162465392e56116195c2304/rclcpp/src/rclcpp/graph_listener.cpp#L375-L382

Not sure if it's worth the effort for a couple of lines though. Might revisit this later if necessary, but I think this PR can be merged as is.

Blast545 commented:
Rerunning CI after removing two tests with problems. Opened an issue for each one to track them.

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=12504)](http://ci.ros2.org/job/ci_linux/12504/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7472)](http://ci.ros2.org/job/ci_linux-aarch64/7472/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10216)](http://ci.ros2.org/job/ci_osx/10216/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=12424)](http://ci.ros2.org/job/ci_windows/12424/)

nuclearsandwich commented:
@Blast545 @brawner one of the tests introduced in this PR is failing on Windows Debug https://github.com/ros2/rclcpp/issues/1391
